# SQL проект

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Test', # пароль
'host': 'mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])


In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [4]:
query = '''
SELECT *
FROM books
'''
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [5]:
query = '''
SELECT *
FROM authors
'''
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [6]:
query = '''
SELECT *
FROM publishers
'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [7]:
query = '''
SELECT *
FROM ratings
'''
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [8]:
query = '''
SELECT *
FROM reviews
'''
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

Посчитайте, сколько книг вышло после 1 января 2000 года;

In [9]:
query = '''
SELECT count(book_id)
FROM books
where publication_date >= '2000-01-02'
'''
pd.io.sql.read_sql(query, con = engine)

count
0    819

Всего 819 книг

Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [10]:
query = '''
WITH  rev as(SELECT title,books.book_id ,count(text)
FROM books
left join reviews  on books.book_id=reviews.book_id 
group by books.book_id),

rat as (SELECT title,books.book_id ,avg(rating)
FROM books
left join ratings  on books.book_id=ratings.book_id
group by books.book_id)

SELECT  rev.title,rev.book_id ,count,avg
from rev
left  join rat on rev.book_id=rat.book_id
order by count DESC,avg DESC

'''
pd.io.sql.read_sql(query, con = engine)
 

title  book_id  count  \
0                              Twilight (Twilight  #1)      948      7   
1    Harry Potter and the Prisoner of Azkaban (Harr...      302      6   
2    Harry Potter and the Chamber of Secrets (Harry...      299      6   
3                                       The Book Thief      656      6   
4                                     The Glass Castle      734      6   
..                                                 ...      ...    ...   
995  Disney's Beauty and the Beast (A Little Golden...      191      0   
996                               Leonardo's Notebooks      387      0   
997                          Essential Tales and Poems      221      0   
998    Anne Rice's The Vampire Lestat: A Graphic Novel       83      0   
999                            The Natural Way to Draw      808      0   

          avg  
0    3.662500  
1    4.414634  
2    4.287500  
3    4.264151  
4    4.206897  
..        ...  
995  4.000000  
996  4.000000  
997  4.000000  
998  3.666667  
999  3.000000  

[1000 rows x 4 columns]

Лидируют Twilight (Twilight #1), The Hobbit or There and Back Again	и The Catcher in the Rye.

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [11]:
query = '''
SELECT publishers.publisher_id ,publisher,count(book_id) AS count
FROM publishers
inner join books  on publishers.publisher_id=books.publisher_id 
where num_pages >50
group by publishers.publisher_id
order by count desc
'''
pd.io.sql.read_sql(query, con = engine)

publisher_id                     publisher  count
0             212                 Penguin Books     42
1             309                       Vintage     31
2             116      Grand Central Publishing     25
3             217              Penguin Classics     24
4              33              Ballantine Books     19
..            ...                           ...    ...
329            34         Ballantine Books (NY)      1
330           225                 Plaza y Janés      1
331           138       HarperCollinsPublishers      1
332           245  Random House: Modern Library      1
333           205                 Pan Childrens      1

[334 rows x 3 columns]

Penguin Books - явный лидер. Издательство выпустило 42 книги.

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [12]:
query = '''
SELECT  author,avg(avg)
from (SELECT  author,title,books.book_id,avg(rating),count(ratings.rating_id)
FROM authors
inner join books  on authors.author_id  =books.author_id   
inner join ratings  on books.book_id =ratings.book_id  
group by author,books.book_id
having count(ratings.rating_id)>50
order by avg(rating) desc) as avgg
group by author
order by avg desc
'''
pd.io.sql.read_sql(query, con = engine)

author       avg
0                          J.K. Rowling/Mary GrandPré  4.283844
1                   Markus Zusak/Cao Xuân Việt Khương  4.264151
2                                      J.R.R. Tolkien  4.258446
3                                   Louisa May Alcott  4.192308
4                                        Rick Riordan  4.080645
5                                     William Golding  3.901408
6                                       J.D. Salinger  3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
9                                           Dan Brown  3.754540
10                                         Lois Lowry  3.750000
11            George Orwell/Boris Grabnar/Peter Škerl  3.729730
12                                    Stephenie Meyer  3.662500
13                                     John Steinbeck  3.622951

Самая высокая средняя оценка - 4.283844 у J.K. Rowling/Mary GrandPré.

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [13]:
query = '''
WITH rat as (SELECT ratings.username,count(rating)
FROM ratings
group by ratings.username
having count(rating)>50
order by count desc),

rev as (SELECT username, COUNT(text) as text_count
FROM reviews
group by username)

select avg(text_count) from rev
inner join rat  using(username)



'''
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

Получили всего 24.3 - среднее количество обзоров от пользователей, которые поставили больше 50 оценок.